Modules/_testinternalcapi/test_critical_sections.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * C Extension module to test pycore_critical_section.h API.
+ */
+
+#include "parts.h"
+
+#include "pycore_critical_section.h"
+
+#ifdef Py_NOGIL
+#define assert_nogil assert
+#define assert_gil(x)
+#else
+#define assert_gil assert
+#define assert_nogil(x)
+#endif
+
+
+static PyObject *
+test_critical_sections(PyObject *self, PyObject *Py_UNUSED(args))
+{
+    PyObject *d1 = PyDict_New();
+    assert(d1 != NULL);
+
+    PyObject *d2 = PyDict_New();
+    assert(d2 != NULL);
+
+    // Beginning a critical section should lock the associated object and
+    // push the critical section onto the thread's stack (in Py_NOGIL builds).
+    Py_BEGIN_CRITICAL_SECTION(d1);
+    assert_nogil(PyMutex_IsLocked(&d1->ob_mutex));
+    assert_nogil(_PyCriticalSection_IsActive(PyThreadState_GET()->critical_section));
+    assert_gil(PyThreadState_GET()->critical_section == 0);
+    Py_END_CRITICAL_SECTION;
+    assert_nogil(!PyMutex_IsLocked(&d1->ob_mutex));
+
+    Py_BEGIN_CRITICAL_SECTION2(d1, d2);
+    assert_nogil(PyMutex_IsLocked(&d1->ob_mutex));
+    assert_nogil(PyMutex_IsLocked(&d2->ob_mutex));
+    Py_END_CRITICAL_SECTION2;
```

Comment:
What about before-and-after?

```suggestion
    assert_nogil(!PyMutex_IsLocked(&d1->ob_mutex));
    assert_nogil(!PyMutex_IsLocked(&d2->ob_mutex));
    Py_BEGIN_CRITICAL_SECTION2(d1, d2);
    assert_nogil(PyMutex_IsLocked(&d1->ob_mutex));
    assert_nogil(PyMutex_IsLocked(&d2->ob_mutex));
    Py_END_CRITICAL_SECTION2;
    assert_nogil(!PyMutex_IsLocked(&d1->ob_mutex));
    assert_nogil(!PyMutex_IsLocked(&d2->ob_mutex));
```